File: android/SceneWorld001/app/src/main/java/z3d/vo/SkinMesh.java
package z3d.vo;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import z3d.base.CallBackFun;
import z3d.base.MeshData;
import z3d.base.Scene_data;
import z3d.engine.ResCount;
import z3d.material.Material;
import z3d.material.MaterialBackFun;
import z3d.material.MaterialBaseParam;
import z3d.material.MaterialManager;
import z3d.program.MaterialAnimShader;
import z3d.program.MaterialShader;
import z3d.units.AnimManager;

public class SkinMesh extends ResCount {

    public List<MeshData> meshAry ;
    public List<Matrix3D> bindPosMatrixAry;
    public List<Matrix3D> bindPosInvertMatrixAry;
    public HashMap boneSocketDic;
    public float fileScale ;
    public float tittleHeight ;
    public Vector2D hitBox ;
    public int type ;
    public HashMap animDic ;
    public boolean ready ;
    public List<String>  animUrlAry ;
    public List<List<Float> >  lightData ;
    public List<Vector3D> hitPosItem;
    public HashMap allParticleDic;
    public String url;
    public boolean hasDestory ;
    public SkinMesh()
    {
        super();
        this.animDic=new HashMap();
        this.meshAry=new ArrayList<>();
    }
    public void makeHitBoxItem()
    {

    }
    public void addMesh(MeshData $mesh)
    {
        $mesh.uid = this.meshAry.size();
        this.meshAry.add($mesh);
    }
    public  void  loadMaterial(){
        for (int i = 0; i < this.meshAry.size(); i++){
            this.loadByteMeshDataMaterial(meshAry.get(i),null);
        }
    }

    private void loadByteMeshDataMaterial(MeshData meshData, CallBackFun backFun) {

        MaterialManager.getInstance().getMaterialByte(meshData.materialUrl, new MaterialBackFun() {
            @Override
            public void Bfun(Material value) {
                meshData.material = value;
                meshData.materialParam = new MaterialBaseParam();
                meshData.materialParam.setData(value, meshData.materialParamData);
            }
        },true, MaterialAnimShader.shaderNameStr, new MaterialAnimShader());

    }

    public  void  setAction(List<String> actionAry,String roleUrl )
    {
        this.animUrlAry=new ArrayList<>();
        for (int i = 0; i < actionAry.size(); i++) {
            String name  = actionAry.get(i);
            String url = roleUrl +name;
            AnimData anim=  AnimManager.getInstance().getAnimDataImmediate(url);
            anim.processMesh(this);
            this.animDic.put(name,anim);
            this.animUrlAry.add(url);
        }
    }


}

File: listios/StorageStarted/pan3d/display/role/vo/SkinMesh.m
//
//  SkinMesh.m
//  iosgl
//
//  Created by zhao on 2/4/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "SkinMesh.h"
#import "Matrix3D.h"
#import "GL_Header.h"
#import "MeshData.h"
#import "Vector2D.h"
#import "MaterialBaseParam.h"
#import "MaterialAnimShader.h"
#import "MaterialManager.h"
#import "AnimManager.h"
#import "Scene_data.h"
@interface SkinMesh()


@property(nonatomic,strong)NSMutableArray<NSString*>*   animUrlAry;
@property(nonatomic,strong)NSMutableArray<NSNumber*>*   lightData;
@property(nonatomic,strong)NSMutableArray<Vector3D*>*   hitPosItem;

@property(nonatomic,assign)float  type;

@end
@implementation SkinMesh
 
- (instancetype)init:(Scene3D *)value
{
    self = [super init:value];
    if (self) {
        self.animDic=[[NSMutableDictionary alloc]init];
          self.meshAry=[[NSMutableArray alloc]init];
    }
    return self;
}
-(void)makeHitBoxItem;
{
    
}
-(void)addMesh:(MeshData*)mesh;
{
    mesh.uid=(int)self.meshAry.count;
    [self.meshAry addObject:mesh];
}
-(void)loadMaterial;
{
    [self loadMaterial:^(NSString *localPath) {
        
        
    }];
}
-(void)loadMaterial:(SuccessBlock)fun;
{
    for (int i = 0; i < self.meshAry.count; i++){
     
         [self loadByteMeshDataMaterial:self.meshAry[i] fun:fun];
        }
}
-(void)loadByteMeshDataMaterial:(MeshData*)meshData fun:(SuccessBlock)fun;
{
 
    NSString* url =meshData.materialUrl;
        url= [url stringByReplacingOccurrencesOfString:@"_byte.txt" withString:@".txt"];
        url= [url stringByReplacingOccurrencesOfString:@".txt" withString:@"_byte.txt"];
  
    [self.scene3D.materialManager getMaterialByte:[[Scene_data default]getWorkUrlByFilePath:url] fun:^(NSObject *obj) {
        meshData.material=(Material*)obj;
        if (meshData.materialParamData){
            meshData.materialParam =[[MaterialBaseParam alloc]init:self.scene3D];
           [meshData.materialParam setData:meshData.material ary:meshData.materialParamData];
                           
       }
  
    } info:nil autoReg:YES regName:MaterialAnimShader.shaderStr shader3DCls:[[MaterialAnimShader alloc]init:self.scene3D]];
    
    
     
    
 
}

-(void)setAction:(NSMutableArray<NSString*>*)actionAry roleUrl:(NSString*)roleUrl;
{
   self.animUrlAry =[[NSMutableArray alloc]init];
    for (int i = 0; i < actionAry.count; i++) {
        NSString* name  = actionAry[i];
        NSString* url = [roleUrl stringByAppendingString:actionAry[i]];
        AnimData* anim = [self.scene3D.animManager getAnimDataImmediate:url];
        [anim processMesh:self];
        self.animDic[name] = anim;
        [self.animUrlAry addObject:url];
    }
}
/*
public setAction(actionAry: Array<string>,roleUrl:string): void {
      this.animUrlAry = new Array;
      for (var i: number = 0; i < actionAry.length; i++) {
          var name: string = actionAry[i];
          var url: string = roleUrl + actionAry[i];
          var anim: AnimData = AnimManager.getInstance().getAnimDataImmediate(url);
          anim.processMesh(this);
          this.animDic[name] = anim;
          this.animUrlAry.push(url);
      }
  }
*/
@end
